### PR TITLE
Seperate proposed from active pocs, implement faster activation selection

### PR DESCRIPTION
**Pull request by Vagabond**
- [x] Fix Snapshots
- [x] Fix Purge
- [x] Check what happens if the iterator goes off the end and if that breaks subsequent selections

### CODE REVIEW

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 1]
```diff
@@ -998,6 +1004,7 @@ raw_fingerprint(Ledger, Extended) ->
                        <<"securities_fingerprint">> => SecuritiesHash,
                        <<"routings_fingerprint">> => RoutingsHash,
                        <<"poc_fingerprint">> => PoCsHash,
+                       <<"potential_poc_fingerprint">> => ProposedPoCsHash,
```
Comment (vihu) [inline src/ledger/v1/blockchain_ledger_v1.erl:1007]:
```suggestion
                       <<"proposed_poc_fingerprint">> => ProposedPoCsHash,
```

Comment (vihu) [reply src/ledger/v1/blockchain_ledger_v1.erl:1007]: ^ not necessary though

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 2]
```diff
@@ -2195,6 +2202,15 @@ delete_public_poc(OnionKeyHash, Ledger) ->
     PoCsCF = pocs_cf(Ledger),
     cache_delete(Ledger, PoCsCF, OnionKeyHash).
 
+delete_proposed_poc(OnionKeyHash, Ledger) ->
```
Comment (vihu) [inline src/ledger/v1/blockchain_ledger_v1.erl:2205]:
```suggestion
-spec delete_proposed_poc(OnionKeyHash :: binary(), Ledger :: ledger()) -> ok.
delete_proposed_poc(OnionKeyHash, Ledger) ->
```

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 3]
```diff
@@ -2195,6 +2202,15 @@ delete_public_poc(OnionKeyHash, Ledger) ->
...
+    PoCsCF = proposed_pocs_cf(Ledger),
+    cache_delete(Ledger, PoCsCF, OnionKeyHash).
+
+promote_public_poc(POC, Ledger) ->
```
Comment (vihu) [inline src/ledger/v1/blockchain_ledger_v1.erl:2209]:
```suggestion
-spec promote_public_poc(POC :: blockchain_ledger_poc_v3:poc(), Ledger :: ledger()) -> ok.
promote_public_poc(POC, Ledger) ->
```

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 4]
```diff
@@ -2122,6 +2121,31 @@ process_poc_proposals(BlockHeight, BlockHash, Ledger) ->
...
+    rocksdb:iterator_move(Iter, {seek, <<3, RandHash/binary>>}),
+    NewAcc = case rocksdb:iterator_move(Iter, next) of
+        {ok, _Key, Binary} ->
+            POC = blockchain_ledger_poc_v3:deserialize(Binary),
+            ActivePOC0 = blockchain_ledger_poc_v3:status(active, POC),
+            ActivePOC1 = blockchain_ledger_poc_v3:block_hash(BlockHash, ActivePOC0),
+            ActivePOC2 = blockchain_ledger_poc_v3:start_height(BlockHeight, ActivePOC1),
+            promote_public_poc(ActivePOC2, Ledger),
+            [ActivePOC2 | Acc];
+        {error, _} ->
```
Comment (evanmcc) [inline src/ledger/v1/blockchain_ledger_v1.erl:2140]: we should probably just seek to the beginning and only return early if we see the original key

Comment (evanmcc) [reply src/ledger/v1/blockchain_ledger_v1.erl:2140]: ignore this

**Review by vihu — APPROVED** [review]

**Pull request merged**